lib/logproto/logproto-framed-server.c

```diff
@@ -213,9 +225,103 @@ log_proto_framed_server_fetch(LogProtoServer *s, const guchar **msg, gsize *msg_
               self->buffer_end = self->buffer_end - self->buffer_pos;
               self->buffer_pos = 0;
             }
-          goto read_message;
+          if (self->state == LPFSS_TRIM_MESSAGE)
+            goto trim_message;
+          else
+            goto read_message;
+        }
+      break;
+
+    case LPFSS_TRIM_MESSAGE:
+
+trim_message:
+      if (self->buffer_end == self->buffer_size)
+        {
+          /* The buffer is full */
+          *msg = &self->buffer[self->buffer_pos];
+          *msg_len = self->buffer_end - self->buffer_pos;
+          self->frame_len -= *msg_len;
+          self->state = LPFSS_CONSUME_TRIMMED;
+          self->half_message_in_buffer = TRUE;
+          self->buffer_pos = self->buffer_end = 0;
+          return LPS_SUCCESS;
+        }
+
+      status = log_proto_framed_server_fetch_data(self, may_read, aux);
+      if (status != LPS_SUCCESS)
+        {
+          log_transport_aux_data_reinit(aux);
+          return status;
+        }
+
+      if (self->buffer_end == self->buffer_size)
+        {
+          /* The buffer is full */
+          *msg = &self->buffer[self->buffer_pos];
+          *msg_len = self->buffer_end - self->buffer_pos;
+          self->frame_len -= *msg_len;
+          self->state = LPFSS_CONSUME_TRIMMED;
+          self->half_message_in_buffer = TRUE;
+          self->buffer_pos = self->buffer_end = 0;
+          return LPS_SUCCESS;
         }
       break;
+
+    case LPFSS_CONSUME_TRIMMED:
+      /* Since trimming requires a full (buffer sized) message, the consuming
+       * always starts at the beginning of the buffer, with a new read. */
+consume_trimmed:
+      self->half_message_in_buffer = FALSE;
+      status = log_proto_framed_server_fetch_data(self, may_read, aux);
+      if (status != LPS_SUCCESS)
+        {
+          log_transport_aux_data_reinit(aux);
+          return status;
+        }
+
+      // Fetch data can not report EAGAIN, this means no data in the buffer.
+      if (self->buffer_pos == self->buffer_end)
+        return LPS_SUCCESS;
+
+      // We have some data, but it is not enough.
+      if (self->buffer_end < self->frame_len)
+        {
+          self->frame_len -= self->buffer_end;
+          buffer_was_full = self->buffer_end == self->buffer_size;
+          self->buffer_end = 0;
+
+          if (buffer_was_full)
+            goto consume_trimmed;
+          return LPS_SUCCESS;
+        }
+
+      // We have enough data to finish consuming the message.
+      self->buffer_pos += self->frame_len;
+      self->state = LPFSS_FRAME_READ;
+
+      // Prepare the buffer for the new message.
+
+      // No more data in the buffer
+      if (self->buffer_pos == self->buffer_end)
```

Comment:
Actually both:
1. the `self->buffer_pos == self->buffer_end` is an optimalization. (Whenever the buffer became empty, it is a chance to "jump" to the beginning without any costly memmove operation later.)
2. It has a functional purpose in a way, that if this "no data" scenario occurs at the end of the buffer, than we can assume that there MIGHT be data in the socket buffer, waiting to be read. We just didn't read it yet, because our buffer was full.